lib/tools/node-modules/node-modules-dependencies-builder.ts

```diff
@@ -1,111 +1,98 @@
 
-export class NodeModulesDependenciesBuilder implements INodeModulesDependenciesBuilder {
-	private projectPath: string;
-	private rootNodeModulesPath: string;
-	private resolvedDependencies: any[];
-	private seen: any;
-
-	public constructor(private $fs: IFileSystem) {
-		this.seen = {};
-		this.resolvedDependencies = [];
-	}
-
-	public getProductionDependencies(projectPath: string): any[] {
-		this.projectPath = projectPath;
-		this.rootNodeModulesPath = path.join(this.projectPath, "node_modules");
-
-		let projectPackageJsonpath = path.join(this.projectPath, "package.json");
-		let packageJsonContent = this.$fs.readJson(projectPackageJsonpath);
-
-		_.keys(packageJsonContent.dependencies).forEach(dependencyName => {
-			let depth = 0;
-			let directory = path.join(this.rootNodeModulesPath, dependencyName);
+interface IDependencyDescription {
+	parentDir: string;
+	name: string;
+	depth: number;
+}
 
-			// find and traverse child with name `key`, parent's directory -> dep.directory
-			this.traverseDependency(dependencyName, directory, depth);
-		});
+export class NodeModulesDependenciesBuilder implements INodeModulesDependenciesBuilder {
+	public constructor(private $fs: IFileSystem) { }
+
+	public getProductionDependencies(projectPath: string): IDependencyData[] {
+		const rootNodeModulesPath = path.join(projectPath, NODE_MODULES_FOLDER_NAME);
+		const projectPackageJsonPath = path.join(projectPath, PACKAGE_JSON_FILE_NAME);
+		const packageJsonContent = this.$fs.readJson(projectPackageJsonPath);
+		const dependencies = packageJsonContent && packageJsonContent.dependencies;
+
+		let resolvedDependencies: IDependencyData[] = [];
+
+		let queue: IDependencyDescription[] = _.keys(dependencies)
+			.map(dependencyName => {
+				return {
+					parentDir: projectPath,
+					name: dependencyName,
+					depth: 0
+				};
+			});
 
-		return this.resolvedDependencies;
-	}
+		while (queue.length) {
+			const currentModule = queue.shift();
+			const resolvedDependency = this.findModule(rootNodeModulesPath, currentModule.parentDir, currentModule.name, currentModule.depth);
 
-	private traverseDependency(name: string, currentModulePath: string, depth: number): void {
-		// Check if child has been extracted in the parent's node modules, AND THEN in `node_modules`
-		// Slower, but prevents copying wrong versions if multiple of the same module are installed
-		// Will also prevent copying project's devDependency's version if current module depends on another version
-		let modulePath = path.join(currentModulePath, "node_modules", name); // node_modules/parent/node_modules/<package>
-		let alternativeModulePath = path.join(this.rootNodeModulesPath, name);
+			if (resolvedDependency && !_.some(resolvedDependencies, r => r.directory === resolvedDependency.directory)) {
+				const deps = _.map(resolvedDependency.dependencies, d => ({ name: d, parentDir: resolvedDependency.directory, depth: resolvedDependency.depth + 1 }));
```

Comment:
I think we need to filter out here for:

1. Not add to the Queue dependencies we've already resolved (i.e. if 2 modules depend on lodash, we'll add both for traversing)
2. If we have Cyclic reference in the Dependencies, we can add them forever as we don't check when adding to the queue.